### PR TITLE
dashboard: archive of workflow

### PR DESCRIPTION
**Pull request by STRRL**
Signed-off-by: STRRL <str_ruiling@outlook.com>

### What problem does this PR solve?

Implement archive API for workflow. 

### What is changed and how it works?

What's Changed:

- use `gorm` for persistent Workflow
- related `/archive/workflows` dashboard APIs
- parameter changes on `/workflows` dashboard APIs

### Related changes

* PR to update `chaos-mesh/website`/`chaos-mesh/website-zh`: No.
* Need to update Chaos Dashboard component, related issue: Yes, no issue yet.
* Need to cheery-pick to the release branch: No.

### Checklist

Tests

- [ ] Unit test
- [ ] E2E test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Breaking backward compatibility

### Release note 

```release-note
Please add a release note.
If you don't think this PR needs a release note then fill it with None.
```

### CODE REVIEW

**Comment by cwen0** [comment]
@STRRL CI is broken. 

```
2021/06/10 11:44:56 ParseComment error in file pkg/apiserver/workflow/workflow.go :can not find type def: "core.Workflow" error: type spec not found
make: *** [Makefile:107: swagger_spec] Error 1
```

**Comment by STRRL** [comment]
> core.Workflow

Fixed! PTAL!

File: pkg/apiserver/archive/archive.go [hunk 1]
```diff
@@ -472,3 +480,181 @@ func (s *Service) batchDeleteSchedule(c *gin.Context) {
...
+			c.Status(http.StatusInternalServerError)
+			_ = c.Error(utils.ErrInvalidRequest.New("the archive schedule is not found"))
+		} else {
+			c.Status(http.StatusInternalServerError)
+			_ = c.Error(utils.ErrInternalServer.NewWithNoMessage())
+		}
+		return
+	}
+
+	sch := &v1alpha1.Workflow{}
```
Comment (fewdan) [inline pkg/apiserver/archive/archive.go:558]: Should "sch" be renamed? Change to "workflow" related?

Comment (STRRL) [reply pkg/apiserver/archive/archive.go]: fixed! PTAL!

File: pkg/apiserver/archive/archive.go [hunk 2]
```diff
@@ -472,3 +480,181 @@ func (s *Service) batchDeleteSchedule(c *gin.Context) {
...
+		_ = c.Error(utils.ErrInternalServer.NewWithNoMessage())
+		return
+	}
+
+	archives := make([]Archive, 0)
+
+	for _, meta := range metas {
+		parsedTime, err := time.Parse(time.RFC3339, meta.CreatedAt)
+		if err != nil {
+			c.Status(http.StatusInternalServerError)
```
[the remark this answers is not in the document]
Comment (g1eny0ung) [reply pkg/apiserver/archive/archive.go:509]: The error will exist in the response. So I think it's no need to print.

File: pkg/collector/workflow_collector.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+func (it *WorkflowCollector) persistentWorkflow(workflow *v1alpha1.Workflow) error {
+	newEntity, err := core.WorkflowCR2WorkflowEntity(workflow)
+	if err != nil {
+		return err
+	}
+
+	existedEntity, err := it.store.FindByUID(context.Background(), string(workflow.UID))
+	if err != nil && !gorm.IsRecordNotFoundError(err) {
+		it.Log.Error(err, "failed to find schedule", "UID", workflow.UID)
```
Comment (g1eny0ung) [inline pkg/collector/workflow_collector.go:91]: Should check all logs. It is ok to reuse logic, but the critical log info should also be modified at the same time.

File: pkg/apiserver/archive/archive.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cwen0) [reply pkg/apiserver/archive/archive.go:509]: I think the error message should be included in the log. This is an important message for debugging.

[the remark this answers is not in the document]
Comment (STRRL) [reply pkg/apiserver/archive/archive.go:509]: There are no logs in several APIs of the dashbaord, `/event` and `/archive` . I think it's better to add these in another PR.

File: pkg/collector/workflow_collector.go (hunk 3 above)
Comment (STRRL) [reply pkg/collector/workflow_collector.go]: fixed in 2039599e

**Review by g1eny0ung — APPROVED** [review]

File: pkg/apiserver/workflow/workflow.go [hunk 4]
```diff
@@ -15,8 +15,13 @@ package workflow
 
 import (
 	"encoding/json"
+	"fmt"
 	"net/http"
 
+	ctrl "sigs.k8s.io/controller-runtime"
+
```
[the remark this answers is not in the document]
Comment (STRRL) [reply pkg/apiserver/workflow/workflow.go]: fixed in ff1a4dd7

File: pkg/apiserver/workflow/workflow.go [hunk 5]
```diff
@@ -15,8 +15,13 @@ package workflow
...
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
```
[the remark this answers is not in the document]
Comment (STRRL) [reply pkg/apiserver/workflow/workflow.go]: fixed in ff1a4dd7

File: pkg/apiserver/archive/archive.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (STRRL) [reply pkg/apiserver/archive/archive.go:512]: I have opened a new issue: https://github.com/chaos-mesh/chaos-mesh/issues/1968

File: pkg/apiserver/archive/archive.go [hunk 6]
```diff
@@ -24,18 +24,22 @@ import (
...
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
 
 	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
 	"github.com/chaos-mesh/chaos-mesh/pkg/apiserver/utils"
 	config "github.com/chaos-mesh/chaos-mesh/pkg/config/dashboard"
 	"github.com/chaos-mesh/chaos-mesh/pkg/core"
 )
 
+var log = ctrl.Log.WithName("archive api")
```
[the remark this answers is not in the document]
Comment (STRRL) [reply pkg/apiserver/archive/archive.go:35]: ignore

[the remark this answers is not in the document]
Comment (STRRL) [reply pkg/apiserver/archive/archive.go:35]: ignore

**Review by cwen0 — APPROVED** [review]
LGTM

**Pull request merged**